Add FoamConfig class and support

atgeirr opened the pull request:
This replaces #871, it can be considered my reviewed and updated version of that PR.

Also includes a minor fix for the ParserItem code generation, that bit me during development.

atgeirr commented:
jenkins build this please

atgeirr commented:
Test failure seems due to missing definition for `size_t`, will include `<cstddef>` and change to `std::size_t` to be sure.

atgeirr commented:
jenkins build this please

bska commented on tests/parser/FoamTests.cpp:
```diff
@@ -0,0 +1,109 @@
+
+
+static Deck createDeck() {
+const char *deckData =
```
This really is a case for raw string literals...

joakim-hove commented on src/opm/parser/eclipse/EclipseState/InitConfig/FoamConfig.cpp:
```diff
@@ -0,0 +1,72 @@
+    // FoamConfig member functions.
+
+    FoamConfig::FoamConfig( const DeckKeyword& keyword ) :
+        records( keyword.begin(), keyword.end() )
```
Hmmm - I would have thought this invoked implicit conversion?

atgeirr replied:
Apparently not!

joakim-hove commented on src/opm/parser/eclipse/EclipseState/InitConfig/FoamConfig.cpp:
```diff
@@ -0,0 +1,72 @@
+        records( keyword.begin(), keyword.end() )
+    {}
+
+    const FoamRecord& FoamConfig::getRecord( std::size_t id ) const {
```
The argument is in my opinion an *index* - and not an *id*. Mixing these have led to subtle bugs in opm.

joakim-hove replied:
`getRecord` could be `operator[]`?

joakim-hove commented on src/opm/parser/eclipse/EclipseState/InitConfig/InitConfig.cpp:
```diff
@@ -38,8 +39,14 @@ namespace Opm {
         return Equil( deck.getKeyword<ParserKeywords::EQUIL>(  ) );
     }
 
+    static inline FoamConfig foamconfigs( const Deck& deck ) {
```
Instead of this logic I would let the `FoamConfig` constructor take a `Deck` - possibly as an overload.

atgeirr commented:
I have addressed the comments. I did not change getRecord() to be operator[], since it is a checked call.

A side-note about the implicit conversion or not: when constructing a vector you can pass iterators to acceptable constructor arguments, even if explicit. However when using the vector::assign() method you cannot! This is why the new constructor (taking a Deck) has this line:
```
            this->records = std::vector<FoamRecord>(kw.begin(), kw.end());
```
whereas the alternative
```
            this->records.assign(kw.begin(), kw.end());
```
fails due to missing implicit conversion. I kept the explicit constructor even though it costs one vector move assign operation...

atgeirr commented:
jenkins build this please

atgeirr commented:
> address them if you like - and then merge away!

Issues addressed, jenkins green, no new warnings, merging.
